### ISSUE TITLE
Use conditional mutations in merge and delete range

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#3763]
The metadata algorithms for merging tablets and deleting a range of data from a table need to be update to use conditional mutations. This could possibly be done at the same time as #3650

**Comment by keith-turner** [comment]
This can be done after #3761 is merged into the elasticity branch.

**Comment by DomGarguilo** [comment]
I plan to start looking into this.

**Comment by keith-turner** [comment]
@DomGarguilo the merge of the changes from #3761 were done in  8cd7c59412887b4d06b40610e51412293fca870d.  I made  changes specific to the elasticity branch in that commit to get MergeIT and DeleteRowsIT working.  Although the test are passing, the functionality is not correct relative to other functionality (like compactions) until its adapted to use conditional mutations.

**Comment by keith-turner** [comment]
Split and merge are destructive tablet operations that destroy tablets and create new tablets.  A tablet can not be hosted on a tablet server while these operations happen.  In elasticity, tablet splitcurrently has the ability to unhost tablets and ensure no changes are made to the tablet while the destructive operation is in progress.  Merge could be adapted to use these same mechanisms that exist for splitting a tablet.

In order to ensure that tablets are not modified by anything else, split relies on a new metadata column named operation id.  This column records the fact that an operation is ongoing that requires exclusive write access to the tablet metadata.  Tablet operations like compact, bulk import, tablet loading, etc require that the operation id is not set via a conditional mutation when updating a tablet.

Using operation id, split roughly does the following.

  * Request unassignment of the tablet if its hosted on a tablet server
  * Set an operation id on the tablet, but only if the follow conditions are met.  The code for this is [here](https://github.com/apache/accumulo/blob/7d21ca2cdcad45efc7c606e094a9cbe6c1838112/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/PreSplit.java#L78)
    * Does not have a location
    * Does not have an operation id currently set
  * Make the destructive update to the tablet and create the new tablet with the same operation id.
  * Delete the operation ids from the tablets.

 The merge operation could follow a similar pattern, however it will have one big difference.  Split always starts with a single tablet. A merge operation can start with multiple tablets.  So merge will need to do something like the following.

  * Request unassignment of all tablets in the range.
  * Use conditional mutations to set an operation id on each tablet in the range as long as it does not have a location and does not have an existing operation id. 
  * Adapt the no chop merge changes to use conditional mutations.  These mutations should ensure the operation id is set to the expected value via setting conditions on the mutation.
  * Once all of the metadata updates are complete, delete the operation ids on the tablets.

There is a race condition that could happen if two concurrent merge operations run, what happens if each of them set operations ids on an overlapping set of tablets?  This could cause deadlock if not handled.  However the table locks that FATE operation currently acquire should prevent concurrent merge operations.   It would be follow on work and not part of the initial work in this issue, but if we can devise a metadata algorithm that handles this then we may be able to remove table locks in Accumulo.  The metadata algorithm could do something like if one fate sees another fate has set operation ids, then the fate with the highest id could remove the operation ids it has set and and let the other fate acquire them.

**Comment by keith-turner** [comment]
Discussed this issue with @DomGarguilo  and he will not be able to work on it in the short term, so I am going to take a look at it.
